src/api.ts

```diff
@@ -0,0 +1,73 @@
+export interface Config {
+    url: string | URL;
+    authToken?: string;
+    transactions?: boolean;
+}
+
+export interface Client {
+    execute(stmt: InStatement): Promise<ResultSet>;
+    batch(stmts: Array<InStatement>): Promise<Array<ResultSet>>;
+    transaction(): Promise<Transaction>;
+    close(): void;
+    closed: boolean;
+}
+
+export interface Transaction {
+    execute(stmt: InStatement): Promise<ResultSet>;
+    rollback(): Promise<void>;
+    commit(): Promise<void>;
+    close(): void;
+    closed: boolean;
+}
+
+export interface ResultSet {
+    columns: Array<string>;
+    rows: Array<Row>;
+    rowsAffected: number;
+}
+
+export interface Row {
+    length: number;
+    [index: number]: Value;
+    [name: string]: Value;
+}
+
+export type Value =
+    | null
+    | string
+    | number
+    | ArrayBuffer
+
+export type InValue =
+    | Value
+    | bigint
+    | boolean
+    | Uint8Array
+    | Date
+
+export type InStatement = { sql: string, args: InArgs } | string;
+export type InArgs = Array<InValue> | Record<string, InValue>;
+
+export class LibsqlError extends Error {
+    code: ErrorCode | undefined;
+    
+    constructor(message: string, code: ErrorCode | undefined, cause?: Error) {
+        if (code !== undefined) {
+            message = `${code}: ${message}`;
+        }
+        super(message, { cause });
+        this.code = code;
+    }
+}
+
+export type ErrorCode =
+    | "NOT_IMPLEMENTED"
+    | "URL_SCHEME_NOT_SUPPORTED"
+    | "URL_PARAM_NOT_SUPPORTED"
+    | "URL_PARAM_INVALID_VALUE"
+    | "SERVER_ERROR"
+    | "TRANSACTION_CLOSED"
+    | "CLIENT_CLOSED"
+    | "TRANSACTIONS_DISABLED"
+    | "TRANSACTIONS_NOT_SUPPORTED"
+    | string;
```

Comment:
When can this happen?  Should this case instead have its own value (e.g. "UNKNOWN")?

Reply:
There are two reasons for the `string` fallback:
1. When we implement error codes in the protocol, the codes will be received as arbitrary strings, we won't be able to statically guarantee that only the listed codes will be used.
2. The `string` case keeps the `ErrorCode` type "open ended", so that we can add new variants without breaking backwards compatibility.